### DISCUSSION TITLE
Non-named parameters

### DISCUSSION THREAD

**Discussion by qianxi0410** [typst/typst#1396] [Q&A] [answered]
Do all parameters need to declare a parameter name except the body?

Now I have a function that only needs to accept an array of parameters, how can I pass in the data directly without writing the parameter names?

**Comment by laurmaedje** [comment, marked as answer]
Currently, paramaters with default values always must be named and parameters without default value must be positional:

```
#let f(x, y, z: 3) = x + y + z
#f(1, 2, z: 5) // ok
#f(x: 1, y: 2, z: 5) // not ok, x and y should be positional
#f(1, 2, 3) // not ok, z must be named
#f(1, 2) // ok, z is optional
f() // not ok, x and y are required
```

This is the current state of things. There have been discussions about having the distinction required/optional instead of positional/named and naming would always be optional, but no concrete implementation efforts have been taken yet (there are details to hash out).
